torch/_library/custom_ops.py

```diff
@@ -0,0 +1,363 @@
+    types: device_types_t = None,
+    qualname: Optional[str] = None,
+) -> Callable:
+    """Wraps a function into custom operator.
+
+    Reasons why you may want to create a custom op include:
+    - Wrapping a third-party library or custom kernel to work with PyTorch
+      subsystems like Autograd.
+    - Preventing torch.compile/export/FX tracing from peeking inside your function.
+
+    This API is used as a decorator around a function (please see examples).
+    The provided function must have type hints; these are needed to interface
+    with PyTorch's various subsystems.
+
+    Args:
+        mutated_args (Sequence[str]): The names of args that the function mutates.
+            This MUST be accurate, otherwise, the behavior is undefined.
+        types (None | str | Sequence[str]): The device type(s) the function
+            is valid for. If no device type is provided, then the function
+            is used as the default implementation for all device types.
+            Examples: "cpu", "cuda".
+        qualname (None | str): An optional name for the operator. If provided,
+            must look like "{namespace}::{name}", e.g. "mylib::my_linear"
+
+    Examples::
+        >>> import torch
+        >>> from torch import Tensor
+        >>> from torch.library import custom_op
+        >>> import numpy as np
+        >>>
+        >>> @custom_op(mutated_args=())
+        >>> def numpy_sin(x: Tensor) -> Tensor:
+        >>>     x_np = x.cpu().numpy()
+        >>>     y_np = np.sin(x_np)
+        >>>     return torch.from_numpy(y_np).to(device=x.device)
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = numpy_sin(x)
+        >>> assert torch.allclose(y, x.sin())
+        >>>
+        >>> # Example of a custom op that only works for one device type.
+        >>> @custom_op(mutated_args=(), types="cpu")
+        >>> def numpy_sin_cpu(x: Tensor) -> Tensor:
+        >>>     x_np = x.numpy()
+        >>>     y_np = np.sin(x_np)
+        >>>     return torch.from_numpy(y_np)
+        >>>
+        >>> x = torch.randn(3)
+        >>> y = numpy_sin_cpu(x)
+        >>> assert torch.allclose(y, x.sin())
+
+    """
+    assert len(mutated_args) == 0, "NYI"
+    assert qualname is None, "NYI"
+
+    def inner(fn):
+        return CustomOpDef._from_fn(fn, types, mutated_args)
+
+    return inner
+
+
+class CustomOpDef:
+    def __init__(self, namespace: str, name: str, schema: str, fn: Callable) -> None:
+        # Fields used to interface with the PyTorch dispatcher
+        self._namespace = namespace
+        self._name = name
+        self._qualname = f"{self._namespace}::{self._name}"
+        self._schema = schema
+
+        self._init_fn = fn
+
+        self._backend_fns: Dict[Union[str, None], Callable] = {}
+        self._abstract_fn: Optional[Callable] = None
+
+        self._lib = get_library_allowing_overwrite(self._namespace, self._name)
+        self._build()
+
+    @classmethod
+    def _from_fn(
```

Comment:
I don't want users to rely on this

Reply:
But they can rely on the rest of the class here?

Reply:
Yup

Reply:
I moved this out to avoid confusion. but in general there are a lot of private attributes on this class (all prefixed with _) that I don't want people to rely on.